### ISSUE TITLE
Weird behavior with decorators

### ISSUE THREAD

**Issue by algmyr** [microsoft/pyright#5146]
**Describe the bug/To Reproduce**
pyright seems to interact weird with decorators. In this example we have two decorators that should be equivalent, and on inspection (through LSP) the type reported for `g` is the same. But the behavior at the end of it all is different.
```py
def deco_v1(_):
    def g():
        return 
    return g

def deco_v2(_):
    def g() -> None:
        return
    return g

@deco_v1
def f1(x):
    return x

@deco_v2
def f2(x):
    return x

f1()  # Argument missing for parameter "x"
f2()  # ok
```

FWIW the thing that triggered investigation in the first place is a similar example
```py
def call(f):
    return f()

@call
def f():
    return lambda x: x

print(f(42))  # Expected 0 positional arguments
```

**Expected behavior**
Neither `f1` nor `f2` should produce an error.

**VS Code extension or command-line**
Using pyright in nvim through LSP, pyright version 1.1.309.

**Comment by erictraut** [comment]
Pyright is working as designed here, so I don't consider this a bug. I can see how this would appear to be inconsistent, but there's a reasonable explanation for what you're seeing.

By default, pyright assumes that unannotated decorators do not modify the signature of the function they're decorating. If your intent is to transform the signature of the decorated function, you should annotate your decorator accordingly. 

Pyright will apply an unannotated decorator if it determines that the return type contains some type information supplied in the form of an annotation. Your `deco_v2` meets this criteria because the `None` return type was explicitly annotated. Pyright refrains from applying the decorator if no type annotations are present. We developed and tuned this heuristic based on our experience with unannotated code bases. The vast majority of decorators do not modify the signature of the function they're decorating, so it's usually safe to assume that behavior when the intent is not explicit.

If you would like to suggest a better heuristic, I'm open to adopting it if it we can demonstrate that it produces better behavior, on average, with unannotated code bases.
